mne/io/base.py

```diff
@@ -2562,98 +2535,49 @@ def set_annotations(self, annotations):
+        logger.info(f"Reserving possible split file {reserved_fname.name}")
+        _check_fname(reserved_fname, overwrite)
+        reserved_ctx = _ReservedFilename(reserved_fname)
 
-    # Expand `~` if present
-    fname = str(_check_fname(fname=fname, overwrite=overwrite))
+        prev_fname, next_fname = None, split_fnames[1]
+        logger.info(f"Writing {fpath}")
+        with start_and_end_file(fpath) as fid, reserved_ctx:
+            is_next_split = raw_fid_writer.write(fid, part_idx, prev_fname, next_fname)
+            logger.info(f"Closing {fpath}")
+        if is_next_split:
+            logger.info(f"Renaming BIDS split file {fpath.name}")
+            shutil.move(fpath, dir_path / split_fnames[0])
 
-    base, ext = op.splitext(fname)
-    if part_idx > 0:
-        if split_naming == "neuromag":
-            # insert index in filename
-            use_fname = "%s-%d%s" % (base, part_idx, ext)
-        else:
-            assert split_naming == "bids"
-            use_fname = _construct_bids_filename(base, ext, part_idx)
-    else:
-        use_fname = fname
-    # check for file existence
-    _check_fname(use_fname, overwrite)
-    # reserve our BIDS split fname in case we need to split
-    if split_naming == "bids" and part_idx == 0:
-        # reserve our possible split name
-        reserved_fname = _construct_bids_filename(base, ext, part_idx)
-        logger.info(f"Reserving possible split file {op.basename(reserved_fname)}")
-        _check_fname(reserved_fname, overwrite)
-        ctx = _ReservedFilename(reserved_fname)
-    else:
-        reserved_fname = use_fname
-        ctx = nullcontext()
-    logger.info("Writing %s" % use_fname)
-
-    picks = _picks_to_idx(info, picks, "all", ())
-    with start_and_end_file(use_fname) as fid:
-        cals = _start_writing_raw(
-            fid, info, picks, data_type, reset_range, raw.annotations
-        )
-        with ctx:
-            final_fname = _write_raw_fid(
-                raw,
-                info,
-                picks,
-                fid,
-                cals,
-                part_idx,
-                start,
-                stop,
-                buffer_size,
-                prev_fname,
-                split_size,
-                use_fname,
-                projector,
-                drop_small_buffer,
-                fmt,
-                fname,
-                reserved_fname,
-                data_type,
-                reset_range,
-                split_naming,
-                overwrite=True,  # we've started writing already above
-            )
-    if final_fname != use_fname:
-        assert split_naming == "bids"
-        logger.info(f"Renaming BIDS split file {op.basename(final_fname)}")
-        ctx.remove = False
-        shutil.move(use_fname, final_fname)
-    if part_idx == 0:
-        logger.info("[done]")
-    return final_fname, part_idx
+        part_idx += 1
+
+    while is_next_split:
```

Comment:
`while` is a bit of a risk because you have to be sure it can never go on forever. There is a tricky thing here -- in this loop the way you ensure no infinite loop is to `assert <thing>`. However, `assert` statements are skipped when running in optimized mode with `python -O`, so we try to avoid using them this way. Instead I at least usually use them as "active comments" like in
```
_check_option('thing', thing, ('a', 'b'))
if thing == 'a':
    ...
else:
    assert thing == 'b'
    ...
```
the `assert` can safely be a no-op / ignored because it's really just saying "# we know this has to be 'a" based on how we coded stuff above".

In any case, I'll push a quick commit to change this to a `for/else` loop instead, which should accomplish the same thing but do so more safely

Reply:
Ok, thanks! Good to know.